src/knot_cloud.c

```diff
@@ -120,127 +96,77 @@ static int map_routing_key_to_msg_type(const char *routing_key)
 }
 
 static struct knot_cloud_msg *create_msg(const char *routing_key,
-					 json_object *jso)
+					 const char *json_str)
 {
+	bool has_err;
+
 	struct knot_cloud_msg *msg = l_new(struct knot_cloud_msg, 1);
 
 	msg->type = map_routing_key_to_msg_type(routing_key);
 
-	switch (msg->type) {
-	case UPDATE_MSG:
-		msg->error = NULL;
-		msg->device_id = parser_get_key_str_from_json_obj(jso,
+	has_err = false;
+	if (msg->type == LIST_MSG) {
+		msg->device_id = NULL;
+	} else {
+		msg->device_id = parser_get_key_str_from_json_str(json_str,
 			KNOT_JSON_FIELD_DEVICE_ID);
-		if (!msg->device_id) {
-			l_error("Malformed JSON message");
-			goto err;
-		}
-
-		msg->list = parser_update_to_list(jso);
-		if (!msg->list) {
-			l_error("Malformed JSON message");
-			goto err;
-		}
+		has_err = msg->device_id ? false : true;
+	}
 
-		break;
-	case REQUEST_MSG:
+	if (msg->type == UPDATE_MSG || msg->type == REQUEST_MSG) {
 		msg->error = NULL;
-		msg->device_id = parser_get_key_str_from_json_obj(jso,
-			KNOT_JSON_FIELD_DEVICE_ID);
-		if (!msg->device_id) {
-			l_error("Malformed JSON message");
-			goto err;
+	} else {
+		if (!parser_is_key_str_or_null(json_str,
+				KNOT_JSON_FIELD_ERROR)) {
+			has_err = true;
+		} else {
+			msg->error = parser_get_key_str_from_json_str(json_str,
+				KNOT_JSON_FIELD_ERROR);
 		}
+	}
```

Comment:
I like we avoid code repetition but I think this way we should be more carefully in case of an API change. For example, if it is necessary to add a new message type, we must verify if it can impact or not these verification, instead of focus only on the new message.

Reply:
My only concern is about the change in this function. I understand the reason, but I don't know if the trade-off is interesting.